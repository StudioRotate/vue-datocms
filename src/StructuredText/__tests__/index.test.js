import { mount } from "@vue/test-utils";
import { StructuredText, RenderError, renderRule } from "../index";
import { isHeading } from "datocms-structured-text-utils";

describe("StructuredText", () => {
  beforeEach(() => {
    jest.spyOn(global.console, "error").mockImplementation(() => {});
  });

  describe("with no value", () => {
    it("renders null", () => {
      const wrapper = mount(StructuredText, {
        propsData: { structuredText: null },
      });
      // await wrapper.vm.$nextTick();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("simple dast with no links/blocks", () => {
    const structuredText = {
      value: {
        schema: "dast",
        document: {
          type: "root",
          children: [
            {
              type: "heading",
              level: 1,
              children: [
                {
                  type: "span",
                  value: "This\nis a\ntitle!",
                },
              ],
            },
          ],
        },
      },
    };

    describe("with default rules", () => {
      it("renders the document", () => {
        const wrapper = mount(StructuredText, {
          propsData: { data: structuredText },
        });
        // await wrapper.vm.$nextTick();
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe("with custom rules", () => {
      it("renders the document", () => {
        const wrapper = mount(StructuredText, {
          structuredText,
          propsData: {
            renderText: (text, key) => {
              return (
                <React.Fragment key={key}>
                  {text.replace(/This/, "That")}!!
                </React.Fragment>
              );
            },
            customRules: [
              renderRule(
                isHeading,
                ({ adapter: { renderNode }, node, children, key }) => {
                  return renderNode(`h${node.level + 1}`, { key }, children);
                },
              ),
            ],
          },
        });
        // await wrapper.vm.$nextTick();

        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe("with links/blocks", () => {
    const structuredText = {
      value: {
        schema: "dast",
        document: {
          type: "root",
          children: [
            {
              type: "heading",
              level: 1,
              children: [
                {
                  type: "span",
                  value: "This is a",
                },
                {
                  type: "span",
                  marks: ["highlight"],
                  value: "title",
                },
                {
                  type: "inlineItem",
                  item: "123",
                },
                {
                  type: "itemLink",
                  item: "123",
                  meta: [{ id: "target", value: "_blank" }],
                  children: [{ type: "span", value: "here!" }],
                },
              ],
            },
            {
              type: "block",
              item: "456",
            },
          ],
        },
      },
      blocks: [
        {
          id: "456",
          __typename: "QuoteRecord",
          quote: "Foo bar.",
          author: "Mark Smith",
        },
      ],
      links: [
        {
          id: "123",
          __typename: "DocPageRecord",
          title: "How to code",
          slug: "how-to-code",
        },
      ],
    };

    describe("with default rules", () => {
      it("renders the document", async () => {
        const wrapper = mount(StructuredText, {
          propsData: {
            data: structuredText,
            renderInlineRecord: ({ record, h, key }) => {
              switch (record.__typename) {
                case "DocPageRecord":
                  return h(
                    "a",
                    {
                      attrs: {
                        href: `/docs/${record.slug}`,
                      },
                      key,
                    },
                    record.title,
                  );
                default:
                  return null;
              }
            },
            renderLinkToRecord: ({
              record,
              children,
              h,
              key,
              transformedMeta,
            }) => {
              switch (record.__typename) {
                case "DocPageRecord":
                  return h(
                    "a",
                    {
                      attrs: {
                        ...transformedMeta,
                        href: `/docs/${record.slug}`,
                      },
                      key,
                    },
                    children,
                  );
                default:
                  return null;
              }
            },
            renderBlock: ({ record, key, h }) => {
              switch (record.__typename) {
                case "QuoteRecord":
                  return h("figure", { key }, [
                    h("blockquote", {}, record.quote),
                    h("figcaption", {}, record.author),
                  ]);
                default:
                  return null;
              }
            },
          },
        });
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe("with missing renderInlineRecord prop", () => {
      it("raises an error", () => {
        expect(() => {
          mount(StructuredText, {
            propsData: { data: structuredText },
          });
        }).toThrow(RenderError);
      });
    });

    describe("with missing record", () => {
      it("raises an error", () => {
        expect(() => {
          mount(StructuredText, {
            propsData: {
              data: { ...structuredText, links: [] },
              renderInlineRecord: () => {
                return null;
              },
            },
          });
        }).toThrow(RenderError);
      });
    });
  });
});
