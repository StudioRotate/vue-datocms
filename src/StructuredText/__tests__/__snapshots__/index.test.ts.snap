// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StructuredText simple dast /2 with default rules renders the document 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": null,
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div>
        <h1>
          This
          <br>
            <!---->
          </br>
          is a
          <br>
            <!---->
          </br>
          title!
        </h1>
        <pre
          data-language="javascript"
        >
          <code>
            function greetings() {
  console.log('Hi!');
}
          </code>
        </pre>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 1,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.11",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "customRules": undefined,
    "data": Object {
      "document": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "type": "span",
                "value": "This
is a
title!",
              },
            ],
            "level": 1,
            "type": "heading",
          },
          Object {
            "code": "function greetings() {
  console.log('Hi!');
}",
            "highlight": Array [
              1,
            ],
            "language": "javascript",
            "type": "code",
          },
        ],
        "type": "root",
      },
      "schema": "dast",
    },
    "hasOwnProperty": [Function],
    "metaTransformer": undefined,
    "renderBlock": undefined,
    "renderFragment": undefined,
    "renderInlineRecord": undefined,
    "renderLinkToRecord": undefined,
    "renderNode": undefined,
    "renderText": undefined,
  },
  "rootVM": Object {},
  "wrapperElement": <div>
    <h1>
      This
      <br>
        <!---->
      </br>
      is a
      <br>
        <!---->
      </br>
      title!
    </h1>
    <pre
      data-language="javascript"
    >
      <code>
        function greetings() {
  console.log('Hi!');
}
      </code>
    </pre>
  </div>,
}
`;

exports[`StructuredText simple dast with no links/blocks with custom rules renders the document 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": null,
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div>
        <h2>
          That!!
          <br>
            <!---->
          </br>
          is a!!
          <br>
            <!---->
          </br>
          title!!!
        </h2>
        <pre
          data-language="javascript"
        >
          <code>
            function greetings() {
  console.log('Hi!');
}!!
          </code>
        </pre>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 3,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.11",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "customRules": Array [
      Object {
        "appliable": [Function],
        "apply": [Function],
      },
    ],
    "data": Object {
      "value": Object {
        "document": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "span",
                  "value": "This
is a
title!",
                },
              ],
              "level": 1,
              "type": "heading",
            },
            Object {
              "code": "function greetings() {
  console.log('Hi!');
}",
              "highlight": Array [
                1,
              ],
              "language": "javascript",
              "type": "code",
            },
          ],
          "type": "root",
        },
        "schema": "dast",
      },
    },
    "hasOwnProperty": [Function],
    "metaTransformer": undefined,
    "renderBlock": undefined,
    "renderFragment": undefined,
    "renderInlineRecord": undefined,
    "renderLinkToRecord": undefined,
    "renderNode": undefined,
    "renderText": [Function],
  },
  "rootVM": Object {},
  "wrapperElement": <div>
    <h2>
      That!!
      <br>
        <!---->
      </br>
      is a!!
      <br>
        <!---->
      </br>
      title!!!
    </h2>
    <pre
      data-language="javascript"
    >
      <code>
        function greetings() {
  console.log('Hi!');
}!!
      </code>
    </pre>
  </div>,
}
`;

exports[`StructuredText simple dast with no links/blocks with default rules renders the document 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": null,
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div>
        <h1>
          This
          <br>
            <!---->
          </br>
          is a
          <br>
            <!---->
          </br>
          title!
        </h1>
        <pre
          data-language="javascript"
        >
          <code>
            function greetings() {
  console.log('Hi!');
}
          </code>
        </pre>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 2,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.11",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "customRules": undefined,
    "data": Object {
      "value": Object {
        "document": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "span",
                  "value": "This
is a
title!",
                },
              ],
              "level": 1,
              "type": "heading",
            },
            Object {
              "code": "function greetings() {
  console.log('Hi!');
}",
              "highlight": Array [
                1,
              ],
              "language": "javascript",
              "type": "code",
            },
          ],
          "type": "root",
        },
        "schema": "dast",
      },
    },
    "hasOwnProperty": [Function],
    "metaTransformer": undefined,
    "renderBlock": undefined,
    "renderFragment": undefined,
    "renderInlineRecord": undefined,
    "renderLinkToRecord": undefined,
    "renderNode": undefined,
    "renderText": undefined,
  },
  "rootVM": Object {},
  "wrapperElement": <div>
    <h1>
      This
      <br>
        <!---->
      </br>
      is a
      <br>
        <!---->
      </br>
      title!
    </h1>
    <pre
      data-language="javascript"
    >
      <code>
        function greetings() {
  console.log('Hi!');
}
      </code>
    </pre>
  </div>,
}
`;

exports[`StructuredText with links/blocks with default rules renders the document 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": null,
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div>
        <h1>
          This is a
          <mark>
            title
          </mark>
          <a
            href="/docs/how-to-code"
          >
            How to code
          </a>
          <a
            href="/docs/how-to-code"
            target="_blank"
          >
            here!
          </a>
        </h1>
        <figure>
          <blockquote>
            Foo bar.
          </blockquote>
          <figcaption>
            Mark Smith
          </figcaption>
        </figure>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 4,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.11",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "customRules": undefined,
    "data": Object {
      "blocks": Array [
        Object {
          "__typename": "QuoteRecord",
          "author": "Mark Smith",
          "id": "456",
          "quote": "Foo bar.",
        },
      ],
      "links": Array [
        Object {
          "__typename": "DocPageRecord",
          "id": "123",
          "slug": "how-to-code",
          "title": "How to code",
        },
      ],
      "value": Object {
        "document": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "span",
                  "value": "This is a",
                },
                Object {
                  "marks": Array [
                    "highlight",
                  ],
                  "type": "span",
                  "value": "title",
                },
                Object {
                  "item": "123",
                  "type": "inlineItem",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "span",
                      "value": "here!",
                    },
                  ],
                  "item": "123",
                  "meta": Array [
                    Object {
                      "id": "target",
                      "value": "_blank",
                    },
                  ],
                  "type": "itemLink",
                },
              ],
              "level": 1,
              "type": "heading",
            },
            Object {
              "item": "456",
              "type": "block",
            },
          ],
          "type": "root",
        },
        "schema": "dast",
      },
    },
    "hasOwnProperty": [Function],
    "metaTransformer": undefined,
    "renderBlock": [Function],
    "renderFragment": undefined,
    "renderInlineRecord": [Function],
    "renderLinkToRecord": [Function],
    "renderNode": undefined,
    "renderText": undefined,
  },
  "rootVM": Object {},
  "wrapperElement": <div>
    <h1>
      This is a
      <mark>
        title
      </mark>
      <a
        href="/docs/how-to-code"
      >
        How to code
      </a>
      <a
        href="/docs/how-to-code"
        target="_blank"
      >
        here!
      </a>
    </h1>
    <figure>
      <blockquote>
        Foo bar.
      </blockquote>
      <figcaption>
        Mark Smith
      </figcaption>
    </figure>
  </div>,
}
`;

exports[`StructuredText with no value renders null 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": null,
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <!---->
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.11",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "customRules": undefined,
    "data": null,
    "hasOwnProperty": [Function],
    "metaTransformer": undefined,
    "renderBlock": undefined,
    "renderFragment": undefined,
    "renderInlineRecord": undefined,
    "renderLinkToRecord": undefined,
    "renderNode": undefined,
    "renderText": undefined,
  },
  "rootVM": Object {},
  "wrapperElement": <!---->,
}
`;
